client/core/core.go

```diff
@@ -3559,6 +3561,38 @@ func (c *Core) Login(pw []byte) (*LoginResult, error) {
 	}
 	defer crypter.Close()
 
+	loginResult := func() *LoginResult {
+		dexStats := c.initializeDEXConnections(crypter)
+		notes, err := c.db.NotificationsN(100)
+		if err != nil {
+			c.log.Errorf("Login -> NotificationsN error: %v", err)
+		}
+
+		return &LoginResult{
+			Notifications: notes,
+			DEXes:         dexStats,
+		}
+	}
+
+	c.mtx.RLock()
+	loginWg := c.loginWg
+	c.mtx.RUnlock()
+
+	if loginWg != nil {
+		// User already sent a login request or is already logged in.
+		c.loginWg.Wait()
+		return loginResult(), err
+	}
+
+	// Connecting and loading wallets might take time, we don't want users
+	// sending multiple login requests when wallets are connecting in a
+	// goroutine.
+	c.mtx.Lock()
+	c.loginWg = &sync.WaitGroup{}
+	c.loginWg.Add(1)
+	c.mtx.Unlock()
+	defer c.loginWg.Done()
```

Comment:
I think this defer should also be called inside the lock. Although it is executed later, it is a defer statement and so evaluated immediately. https://go.dev/play/p/hLeBTmEfagk